Panic on nested TableArray

### Discussed in https://github.com/pelletier/go-toml/discussions/671

<div type='discussions-op-text'>

<sup>Originally posted by **hollowaykeanho** November 18, 2021</sup>
For TOML code that has array of data structure:

```toml
[Build]
Name = "publication build"

[[Build.Dependencies]]
Name = "command"
Program = "hugo"
```

And a Go structure:

```go
type _tomlJob struct {
        Dependencies []map[string]interface{}
}

type TOMLParser struct {
        Build     *_tomlJob
}
```
Parsing Structure:
```
        s := struct {
                Build     *_tomlJob
        }{
                Build:     fx.Build,
        }

```



I got a parsing panic as follows:

```
panic: reflect: call of reflect.Value.Set on zero Value

goroutine 1 [running]:
reflect.flag.mustBeExportedSlow(0x199)
        /home/u0/Documents/monteur/.bin/golang/src/reflect/value.go:237 +0xc5
reflect.flag.mustBeExported(...)
        /home/u0/Documents/monteur/.bin/golang/src/reflect/value.go:231
reflect.Value.Set({0x6e38c0, 0xc0000ae088, 0xc0001023b8}, {0x0, 0x0, 0x6bcd40})
        /home/u0/Documents/monteur/.bin/golang/src/reflect/value.go:1900 +0x9f
github.com/pelletier/go-toml/v2.(*decoder).handleArrayTableCollection(0xc0000bf880, {0x1, 0xc0001023b8}, {0x6bcd40, 0xc0000a6280, 0x5})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:340 +0x50e
github.com/pelletier/go-toml/v2.(*decoder).handleKeyPart(0xc0000bf880, {0x1, 0xc0001023b8}, {0x6f0c40, 0xc0000a6270, 0xc0000bf880}, 0xc0000bf698, 0x73c2e0)
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:439 +0x5fb
github.com/pelletier/go-toml/v2.(*decoder).handleArrayTablePart(0xc0000a6270, {0x1, 0xc0001023b8}, {0x6f0c40, 0xc0000a6270, 0xc0000fea00})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:478 +0x6c
github.com/pelletier/go-toml/v2.(*decoder).handleArrayTable(0xc0000bf728, {0x1, 0xc0001023b8}, {0x6f0c40, 0xc0000a6270, 0x78e2a0})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:252 +0x71
github.com/pelletier/go-toml/v2.(*decoder).handleRootExpression(0xc0000bf880, 0xc000102380, {0x6f0c40, 0xc0000a6270, 0x70})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:234 +0xf7
github.com/pelletier/go-toml/v2.(*decoder).fromParser(0xc0000bf880, {0x6f0c40, 0xc0000a6270, 0x99c380})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:192 +0xa9
github.com/pelletier/go-toml/v2.(*decoder).FromParser(0xc0000bf880, {0x6bfe40, 0xc0000a6270})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:177 +0xf8
github.com/pelletier/go-toml/v2.(*Decoder).Decode(0xc0000bf930, {0x6bfe40, 0xc0000a6270})
        /home/u0/Documents/monteur/pkg/monteur/internal/endec/toml/internal/go-toml/unmarshaler.go:104 +0x171
```

If I change `[[Build.Dependencies]]` into `[[Dependencies]]`, everything works fine. I need the `Dependencies` to be under `Build` because there are many jobs like `Build`, `Publish`, ... which have their respective dependencies.

Is this expected?</div>

--- a/unmarshaler.go
+++ b/unmarshaler.go
@@ -345,7 +345,9 @@ func (d *decoder) handleArrayTableCollection(key ast.Iterator, v reflect.Value) 
 		if err != nil {
 			return reflect.Value{}, err
 		}
-		v.Elem().Set(elem)
+		if elem.IsValid() {
+			v.Elem().Set(elem)
+		}
 
 		return v, nil
 	case reflect.Slice:

--- a/unmarshaler_test.go
+++ b/unmarshaler_test.go
@@ -2220,6 +2220,42 @@ func TestIssue666(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestIssue677(t *testing.T) {
+	doc := `
+[Build]
+Name = "publication build"
+
+[[Build.Dependencies]]
+Name = "command"
+Program = "hugo"
+`
+
+	type _tomlJob struct {
+		Dependencies []map[string]interface{}
+	}
+
+	type tomlParser struct {
+		Build *_tomlJob
+	}
+
+	p := tomlParser{}
+
+	err := toml.Unmarshal([]byte(doc), &p)
+	require.NoError(t, err)
+
+	expected := tomlParser{
+		Build: &_tomlJob{
+			Dependencies: []map[string]interface{}{
+				{
+					"Name":    "command",
+					"Program": "hugo",
+				},
+			},
+		},
+	}
+	require.Equal(t, expected, p)
+}
+
 //nolint:funlen
 func TestUnmarshalDecodeErrors(t *testing.T) {
 	examples := []struct {

